File: src/leetCode/Ex95.java
package leetCode;

import java.util.ArrayList;
import java.util.List;

public class Ex95 {
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) {
            val = x;
        }
    }

    public List<TreeNode> generateTrees(int n) {
        ArrayList<TreeNode> trees = new ArrayList<>();
        for (int i = 0;i<n;i++){
            dfs(i+1,n,trees,new TreeNode(i+1));
        }
        return trees;
    }
    private void dfs(int k,int num,ArrayList<TreeNode> res,TreeNode header){
        
//
    }
    private int i = 0;
    private int f(int n ){
        i++;
        if (n<=3) return 1;
        else return f(n-2)+f(n-6)+1;
    }

    public static void main(String[] args) {
        Ex95 t = new Ex95();
        System.out.println(t.f(t.f(9)));
        System.out.println(t.i);
    }
}

File: src/leetCode/ex95.kt
package leetCode

fun generateTrees(n: Int): List<TreeNode?> {
    return dfsTree(1, n)
}

private fun dfsTree(begin: Int, end: Int): MutableList<TreeNode> {
    val res = mutableListOf<TreeNode>()
    if (begin > end) {
        return res
    } else if (begin == end) {
        res.add(TreeNode(begin))
        return res
    }
    (begin..end).forEach {
        val left = dfsTree(begin, it - 1)
        val right = dfsTree(it + 1, end)
        left.forEach { litem ->
            right.forEach { ritem ->
                res.add(TreeNode(it).apply {
                    this.left = litem
                    this.right = ritem
                })
            }
        }
        if (left.isEmpty() && right.isNotEmpty()) {
            right.forEach { ritem ->
                res.add(TreeNode(it).apply { this.right = ritem })
            }
        }
        if (left.isNotEmpty() && right.isEmpty()) {
            left.forEach { litem ->
                res.add(TreeNode(it).apply { this.left = litem })
            }
        }
    }
    return res

}